ddl/column.go

```diff
@@ -693,17 +705,389 @@ func (w *worker) doModifyColumn(
+				return false, nil
+			}
+
+			_, err := w.rowDecoder.DecodeAndEvalRowWithMap(w.sessCtx, handle, rawRow, time.UTC, sysZone, w.rowMap)
+			if err != nil {
+				return false, errors.Trace(errCantDecodeIndex.GenWithStackByArgs(err))
+			}
+
+			if _, ok := w.rowMap[w.newColInfo.ID]; ok {
+				// The column is already added by update or insert statement, skip it.
+				w.cleanRowMap()
+				return true, nil
+			}
+
+			newColVal, err := table.CastValue(w.sessCtx, w.rowMap[w.oldColInfo.ID], w.newColInfo, false, false)
+			// TODO: Consider sql_mode and the error msg(encounter this error check whether to rollback).
+			if err != nil {
+				return false, errors.Trace(err)
+			}
+			w.rowMap[w.newColInfo.ID] = newColVal
+			newColumnIDs := make([]int64, 0, len(w.rowMap))
+			newRow := make([]types.Datum, 0, len(w.rowMap))
+			for colID, val := range w.rowMap {
+				newColumnIDs = append(newColumnIDs, colID)
+				newRow = append(newRow, val)
+			}
+			sctx, rd := w.sessCtx.GetSessionVars().StmtCtx, &w.sessCtx.GetSessionVars().RowEncoder
+			newRowVal, err := tablecodec.EncodeRow(sctx, newRow, newColumnIDs, nil, nil, rd)
+			if err != nil {
+				return false, errors.Trace(err)
+			}
+
+			w.rowRecords = append(w.rowRecords, &rowRecord{handle: handle, key: recordKey, vals: newRowVal})
+			w.cleanRowMap()
+			if handle.Equal(taskRange.endHandle) {
+				// If taskRange.endIncluded == false, we will not reach here when handle == taskRange.endHandle
+				taskDone = true
+				return false, nil
+			}
+			return true, nil
+		})
+
+	if len(w.rowRecords) == 0 {
+		taskDone = true
 	}
 
+	logutil.BgLogger().Debug("[ddl] txn fetches handle info", zap.Uint64("txnStartTS", txn.StartTS()), zap.String("taskRange", taskRange.String()), zap.Duration("takeTime", time.Since(startTime)))
+	return w.rowRecords, w.getNextHandle(taskRange, taskDone), taskDone, errors.Trace(err)
+}
+
+func (w *updateColumnWorker) cleanRowMap() {
+	for id := range w.rowMap {
+		delete(w.rowMap, id)
+	}
+}
+
+// BackfillDataInTxn will backfill the table record in a transaction, lock corresponding rowKey, if the value of rowKey is changed.
+func (w *updateColumnWorker) BackfillDataInTxn(handleRange reorgIndexTask) (taskCtx backfillTaskContext, errInTxn error) {
+	oprStartTime := time.Now()
+	errInTxn = kv.RunInNewTxn(w.sessCtx.GetStore(), true, func(txn kv.Transaction) error {
+		taskCtx.addedCount = 0
+		taskCtx.scanCount = 0
+		txn.SetOption(kv.Priority, w.priority)
+
+		rowRecords, nextHandle, taskDone, err := w.fetchRowColVals(txn, handleRange)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		taskCtx.nextHandle = nextHandle
+		taskCtx.done = taskDone
+
+		for _, rowRecord := range rowRecords {
+			taskCtx.scanCount++
+
+			err = txn.Set(rowRecord.key, rowRecord.vals)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			taskCtx.addedCount++
+		}
+
+		return nil
+	})
+	logSlowOperations(time.Since(oprStartTime), "BackfillDataInTxn", 3000)
```

Comment:
Why not use `0` as the threshold?

Reply:
This value is set to be the same as “add index”.